### PR TITLE
Submodule for transpose

### PR DESCRIPTION
**Pull request by pbartholomew08**
Reopening the original pull request #204

### CODE REVIEW

**Comment by pbartholomew08** [comment]
We need to hold off merging until we can test on a Cray compiler. I'm 99% sure the Cray compilers refuse to compile code with submodule and parent module both using the same module as in commit 40b0778

**Comment by rfj82982** [comment]
I have a problem with the following line

**Comment by rfj82982** [comment]
The code does not compile with NVHPC. 
```
[  4%] Building Fortran object src/CMakeFiles/decomp2d.dir/transpose_x_to_y.f90.o
cd /work4/scd/scarf924/TrainingGPU/2decomp-fft/build_nvhpc_cpu/src && /work4/scd/scarf562/eb-amd/software/NVHPC/22.7/Linux_x86_64/22.7/comm_libs/mpi/bin/mpif90 -DDOUBLE_PREC -DNVHPC -DVERSION=\"25320a5\" -I/work4/scd/scarf924/TrainingGPU/2decomp-fft/src -I/work4/scd/scarf924/TrainingGPU/2decomp-fft/build_nvhpc_cpu/include -cpp -Mfree -Kieee -O3 -fast -march=native -module ../include -c /work4/scd/scarf924/TrainingGPU/2decomp-fft/src/transpose_x_to_y.f90 -o CMakeFiles/decomp2d.dir/transpose_x_to_y.f90.o
nvfortran-Warning-CUDA_HOME has been deprecated. Please, use NVHPC_CUDA_HOME instead.
NVFORTRAN-S-0038-Symbol, real_type, has not been explicitly declared (/work4/scd/scarf924/TrainingGPU/2decomp-fft/src/transpose_x_to_y.f90)
  0 inform,   0 warnings,   1 severes, 0 fatal for transpose_x_to_y_real
NVFORTRAN-S-0038-Symbol, complex_type, has not been explicitly declared (/work4/scd/scarf924/TrainingGPU/2decomp-fft/src/transpose_x_to_y.f90)
  0 inform,   0 warnings,   1 severes, 0 fatal for transpose_x_to_y_complex
```

**Comment by rfj82982** [comment]
Ok with NVHPC 23.1 is OK the issue is 22.7

**Comment by mathrack** [comment]
@rfj82982 sounds good. I suggest we focus on merging https://github.com/2decomp-fft/2decomp-fft/pull/195 and release 2.0

Version 2.1 will drop compatibility with nvhpc 22.7 and will include this PR ?

**Comment by pbartholomew08** [comment]
N.B. change is much smaller in review if you select the "ignore whitespace changes" option

**Comment by pbartholomew08** [comment]
> We need to hold off merging until we can test on a Cray compiler. I'm 99% sure the Cray compilers refuse to compile code with submodule and parent module both using the same module as in commit [40b0778](https://github.com/2decomp-fft/2decomp-fft/commit/40b0778cd12db53f1ffae15c5f614ee445188121)

Builds under Cray compiler, not AOCC

File: src/CMakeLists.txt
Comment (mathrack) [inline src/CMakeLists.txt:85]: This is still needed ?

**Comment by mathrack** [comment]
This is ready for review and merging in my opinion.

**Comment by rfj82982** [comment]
I'll test this on multi-GPU as soon as I have a bit of time and approve the PR. Of course before I'll merge the new main

**Comment by rfj82982** [comment]
Sorry Dev

**Comment by mathrack** [comment]
> > We need to hold off merging until we can test on a Cray compiler. I'm 99% sure the Cray compilers refuse to compile code with submodule and parent module both using the same module as in commit [40b0778](https://github.com/2decomp-fft/2decomp-fft/commit/40b0778cd12db53f1ffae15c5f614ee445188121)
> 
> Builds under Cray compiler, not AOCC

This PR will probably not compile with AOCC in the foreseeable future. See #308 for instance. It might be best to close without merging ?

**Comment by pbartholomew08** [comment]
> > > We need to hold off merging until we can test on a Cray compiler. I'm 99% sure the Cray compilers refuse to compile code with submodule and parent module both using the same module as in commit [40b0778](https://github.com/2decomp-fft/2decomp-fft/commit/40b0778cd12db53f1ffae15c5f614ee445188121)
> > 
> > 
> > Builds under Cray compiler, not AOCC
> 
> This PR will probably not compile with AOCC in the foreseeable future. See #308 for instance. It might be best to close without merging ? In my opinion, if we want to support aocc (and cray compiler), it would be best to have them in the CI.

We need to buy a Cray to get the Cray compiler ;)

I would be disinclined to hold off on code improvements because flang-based compilers don't compile it - we can still use gfortran, ifort/ifx, nvfortran and crayftn until flang catches up.

**Comment by pbartholomew08** [comment]
Getting Cray compilers into CI isn't really possible (atleast not easily) as they are only available on Cray computers.

**Comment by pbartholomew08** [comment]
And I just checked current `dev`, it doesn't build under AOCC. I think anything based on flang doesn't handle submodules very well, in particular with inheriting variables from the parent module (as far as I can tell this is standards-compliant).

**Comment by mathrack** [comment]
Sounds good. This is ready for merging provided we test and describe the compatibility with nvhpc / CUDA version here https://github.com/2decomp-fft/2decomp-fft/blob/submodule_for_transpose/INSTALL.md?plain=1#L10

**Comment by rfj82982** [comment]
I'll try to add AOCC to our CI machine and see if we want to add to the CI or not. Silly question I have seen AOCC and rocm compilers with clang/flang in it, What are the differences?

**Comment by pbartholomew08** [comment]
My understanding is: AOCC is AMD's compiler collection (like GCC = GNU Compiler Collection). AOCC/flang is their version of (one of) the flang compilers. The issue (as far as I can tell) is somewhere in the flang frontend, because ifx I'm 90% sure is the ifort frontend + LLVM backend - so it's the code generation, rather than LLVM that is the issue.

I think ROCM can be generated from the LLVM intermediate code, but not sure about that.

**Comment by rfj82982** [comment]
I merged the lasted dev. Done the tests also with 4 GPUs and everything meeds to work. I'll confirm do the merge once the CI has completed.

**Comment by mathrack** [comment]
Sounds good. This is ready for merging provided we test and describe the compatibility with nvhpc / CUDA version here https://github.com/2decomp-fft/2decomp-fft/blob/submodule_for_transpose/INSTALL.md?plain=1#L10

Dec 26, 2023 19:00:17 Stefano Rolfo ***@***.***>:

**Review by mathrack — APPROVED** [review]

**Review by rfj82982 — APPROVED** [review]
LGTM

**Pull request merged**
